astropy/units/quantity.py

```diff
@@ -171,10 +171,16 @@ class QuantityInfo(QuantityInfoBase):
     required when the object is used as a mixin column within a table, but can
     be used as a general way to store meta information.
     """
-    _represent_as_dict_attrs = ('value', 'unit')
+    _represent_as_dict_attrs = ('value', 'unit', 'dtype')
     _construct_from_dict_args = ['value']
     _represent_as_dict_primary_data = 'value'
 
+    def _represent_as_dict(self, attrs=None):
+        q = self._parent
+        out = super()._represent_as_dict(attrs)
+        out['dtype'] = q.dtype.name
```

Comment:
This won't work for structured quantities; do we need to overwrite `_represent_as_dict` at all?

I actually wonder why `dtype` is needed at all; in principle, `value` has the dtype information, so that should be propagated correctly if it isn't already...

Reply:
p.s. For the structured types, would be good to include #12509.

Reply:
It needs some instruction on how to use `dtype.name` to store the `dtype` information, apparently. `value` is simply written as something like
`value: !astropy.table.SerializedColumn {name: lon}`
so the `dtype` will have to be passed on to the `Quantity` constructor through some channel.

Reply:
OK, thanks, I now see the problem: the `.value` does have a `dtype` and that gets used, but then the result is passed through `u.Quantity(value, unit)` and the `dtype` is lost at that point. What perhaps we could do is only pass the `dtype` that actually is converted, i.e.,
```
if q.dtype.kind in 'iu':
    out['dtype'] = q.dtype.name
```
This also solves the problem that what you have now would fail for structured dtype (which probably means there should be a test for that...)

Reply:
Yes; I thought about something like that for the first implementation to resolve the `Time` regression either intercepting `object` or other complex dtypes, or restricting explicitly to integers. The latter choice is probably more prudent; this only leaves the corner case of `wrap_angle` in `Longitude`, which is converted to a `float64` default.

Reply:
I think currently scalars are treated differently from arrays, as they are typeset as string instead of a binary blob. Not sure if that is worth changing...

Reply:
> p.s. For the structured types, would be good to include #12509.

@dhomeier @mhvk, merged!